android/app/src/main/java/com/zulipmobile/MainActivity.kt

```diff
@@ -94,28 +96,47 @@ open class MainActivity : ReactActivity() {
         }
     }
 
+    private fun getFileName(uri: Uri, cr: ContentResolver): String? {
+        return cr.query(uri, null, null, null, null)?.use { cursor ->
+            cursor.moveToFirst()
+            val nameIndex = cursor.getColumnIndex(MediaStore.MediaColumns.DISPLAY_NAME)
+            cursor.getString(nameIndex)
```

Comment:
Hmm, I remember having a crash on using `OpenableColumn` several times earlier but this does not happen now for some reason. I used `MediaColumn` as an alternative because I thought that it best represented the type of content sent by a `SEND_INTENT`.

I couldn't find what `OpenableColumn` refers to precisely as well, the [docs](https://developer.android.com/reference/android/provider/OpenableColumns) describe them as columns for "openable uris" but I don't really understand what that refers to.

I have reverted it to `OpenableColumns`.

Reply:
Cool.

> I couldn't find what `OpenableColumn` refers to precisely as well, the [docs](https://developer.android.com/reference/android/provider/OpenableColumns) describe them as columns for "openable uris" but I don't really understand what that refers to.

Hmm, yeah.

It does have a link to this bit of docs:
https://developer.android.com/reference/android/content/Intent#CATEGORY_OPENABLE

I think that contains an implicit answer to the question of what an "openable URI" is supposed to mean:

> Used to indicate that an intent only wants URIs that can be opened with `ContentResolver#openFileDescriptor(Uri, String)`. Openable URIs must support at least the columns defined in OpenableColumns when queried.

It sounds like "openable" here means "can be opened with that openFileDescriptor method".

And then following the link to [that method's doc](https://developer.android.com/reference/android/content/ContentResolver#openFileDescriptor(android.net.Uri,%20java.lang.String)), it accepts the two URL schemes `content:` and `file:`. So I guess only URLs with those two schemes can be "openable". And... also not `content:` URLs where the underlying content provider will return only a part of a file, instead of a whole file.

Ah, but also, following the link from *there* to [`ContentProvider#openFile`](https://developer.android.com/reference/android/content/ContentProvider#openFile(android.net.Uri,%20java.lang.String)):
> For better interoperability with other applications, it is recommended that for any URIs that can be opened, you also support queries on them containing at least the columns specified by OpenableColumns. You may also want to support other common columns if you have additional meta-data to supply, such as MediaStore.MediaColumns.DATE_ADDED in MediaStore.MediaColumns.

So that's something directly about `OpenableColumns` again. That one only "recommends" providers support them. So maybe a way to interpret the OpenableColumns doc:
> Providers that serve openable URIs must support at least these columns when queried.

is... an "openable URI" is one for which the provider supports at least these columns :stuck_out_tongue: 

---

Ah, and in fact the values of these constants are documented -- and `OpenableColumns.DISPLAY_NAME` and `MediaStore.MediaColumns.DISPLAY_NAME` are documented to have the same value, namely `_display_name`.

So both expressions are equivalent.

But `OpenableColumns` is the name used in the upstream docs for this, so it's good to match that.